netsim/extra/node.clone/plugin.py

```diff
@@ -0,0 +1,109 @@
+"""
+def update_links(topo_items: Box, nodename: str, clones: list, topology: Box) -> None:
+  for vname,vdata in topo_items.items():                                    # Iterate over global VLANs or VRFs
+    if not isinstance(vdata,Box):                                           # VLAN not yet a dictionary?
+      continue                                                              # ... no problem, skip it
+    if not 'links' in vdata:                                                # No VLAN links?
+      continue                                                              # ... no problem, move on
+
+    for cnt,l in enumerate(list(vdata.links)):                              # So far so good, now iterate over the links
+      link_data = links.adjust_link_object(                                 # Create link data from link definition
+                    l=l,
+                    linkname=f'vlans.{vname}.links[{cnt+1}]',
+                    nodes=topology.nodes)
+      if link_data is not None:
+        ifs = clone_interfaces(link_data,nodename,clones)
+        if ifs:
+          vdata.links.remove(l)
+          vdata.links += ifs
+
+"""
+clone_node - Clones a given node N times, creating additional links and/or interfaces for the new nodes
+"""
+def clone_node(node: Box, topology: Box) -> None:
+  _p = { 'start': 1, 'step': 1 } + node.pop('clone',{})                      # Define cloning parameters
+  if 'count' not in _p:
+    log.error("Node {node.name} missing required attribute clone.count",     # Not validated by Netlab yet
+              category=AttributeError, module='node.clone')
+    return
+
+  if 'include' in node:                                                      # Check for components
+    log.error("Cannot clone component {node.name}, only elementary nodes",
+              category=AttributeError, module='node.clone')
+    return
+
+  name_format = topology.defaults.clone.node_name_pattern
+
+  orig_name = node.name
+  node.name = strings.eval_format(name_format, node + { 'id': _p.start } )  # Rename first node as clone_01
+  topology.nodes += { node.name: node }
+
+  clones = [ node.name ]
+  for c in range(_p.start+_p.step,_p.start+_p.count*_p.step,_p.step):       # Existing node is '1'
+    clone = data.get_box(node)
+    clone.name = strings.eval_format(name_format, node + { 'id': c, 'name': orig_name } )
+    clone.interfaces = []                                                   # Start clean, remove reference to original node
+    if 'id' in node:
+      clone.id = node.id + c - 1                                            # Update any explicit node ID sequentially
+    topology.nodes += { clone.name: clone }
+    clones.append( clone.name )
+
+  for link in list(topology.get('links',[])):                               # Make a copy of links list
+    link_data = links.adjust_link_object(link,f'clone[{orig_name}]',topology.nodes)
+    if link_data is None:
+      continue
+    if 'lag' in link_data:
```

Comment:
Shouldn't you check this only if the link includes the cloned node? Also, wouldn't it be better to have a common 'link adjustment' function that would work for **links**, **vlans.X.links** and **vrfs.X.links**?

Reply:
If ```'lag' in link_data``` the cloned node could be mentioned in ```lag.members``` - but checking if it's in there amounts to doing the work to support lag links

I was holding off to add ```lag``` support as it's a bit of a moving target, but I can add it. Will build an adjustment function that can also handle ```lag.members```

Reply:
Changing the error message to something like "You cannot use node.clone plugin in a topology with LAG links" might make it easier to understand for non-native speakers like myself. I understood the error message the wrong way.

Reply:
Added lag support and removed the error